File: frontend/src/types/agent.ts
import { Product } from './product';

export interface AgentTrace {
  id: number;
  session_id: string;
  step_number: number;
  phase: 'PLAN' | 'ACT' | 'OBSERVE' | 'APPROVE';
  action_type: string;
  title: string;
  reasoning: string;
  input_data?: Record<string, any>;
  output_data?: Record<string, any>;
  is_reversible: boolean;
  timestamp: string;
}

export interface CandidateComparison {
  product_id: number;
  title: string;
  brand: string;
  price: number;
  original_price?: number;
  rating: number;
  review_count: number;
  score: number;
  specs: Record<string, any>;
  pros: string[];
  cons: string[];
  image_url?: string;
  breakdown: Record<string, string>;
}

export interface TradeOffAnalysis {
  recommended_id?: number;
  recommended_product_title?: string;
  match_score?: number;
  summary_rationale?: string;
  budget_spent?: number;
  budget_remaining?: number;
  candidates?: CandidateComparison[];
}

export interface Order {
  id: string;
  session_id: string;
  product_id: number;
  product_title: string;
  unit_price: number;
  quantity: number;
  total_amount: number;
  shipping_address: string;
  payment_method: string;
  status: string;
  created_at: string;
}

export interface AgentSession {
  id: string;
  user_query: string;
  budget_limit: number;
  status: 'IDLE' | 'PLANNING' | 'SEARCHING' | 'COMPARING' | 'AWAITING_APPROVAL' | 'APPROVED' | 'CHECKING_OUT' | 'COMPLETED' | 'REJECTED' | 'CANCELLED' | 'FAILED';
  current_step: number;
  current_phase: string;
  parsed_intent?: {
    category?: string;
    max_price?: number;
    use_case?: string;
    priorities?: string[];
    target_tags?: string[];
    brand_affinity?: string;
  };
  candidate_product_ids?: number[];
  recommended_product_id?: number;
  recommended_product?: Product;
  candidate_products?: Product[];
  trade_off_analysis?: TradeOffAnalysis;
  cart_item?: {
    product_id: number;
    title: string;
    price: number;
    quantity: number;
    total: number;
    delivery_days: number;
  };
  total_cost: number;
  is_killed: boolean;
  human_feedback?: string;
  created_at: string;
  updated_at: string;
  traces: AgentTrace[];
  order?: Order;
}

File: backend/app/schemas/agent.py
from typing import Optional, List, Dict, Any
from datetime import datetime
from pydantic import BaseModel, Field, ConfigDict
from app.schemas.product import ProductOut

class CreateSessionRequest(BaseModel):
    user_query: str = Field(..., examples=["Find me a budget laptop under $700 for video editing"])
    budget_limit: Optional[float] = Field(1000.0, description="Spending limit cap")

class SessionStepRequest(BaseModel):
    pass

class ApprovalDecisionRequest(BaseModel):
    decision: str = Field(..., pattern="^(APPROVE|REJECT|MODIFY)$", examples=["APPROVE"])
    feedback: Optional[str] = Field(None, description="Optional user notes or instructions if rejected/modified")

class KillSessionRequest(BaseModel):
    reason: Optional[str] = Field("User triggered emergency kill switch", description="Reason for aborting agent task")

class AgentTraceOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    id: int
    session_id: str
    step_number: int
    phase: str # PLAN, ACT, OBSERVE, APPROVE
    action_type: str
    title: str
    reasoning: str
    input_data: Optional[Dict[str, Any]] = None
    output_data: Optional[Dict[str, Any]] = None
    is_reversible: bool
    timestamp: datetime

class OrderOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    id: str
    session_id: str
    product_id: int
    product_title: str
    unit_price: float
    quantity: int
    total_amount: float
    shipping_address: str
    payment_method: str
    status: str
    created_at: datetime

class AgentSessionOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    id: str
    user_query: str
    budget_limit: float
    status: str
    current_step: int
    current_phase: str
    parsed_intent: Optional[Dict[str, Any]] = None
    candidate_product_ids: Optional[List[int]] = None
    recommended_product_id: Optional[int] = None
    recommended_product: Optional[ProductOut] = None
    candidate_products: Optional[List[ProductOut]] = None
    trade_off_analysis: Optional[Dict[str, Any]] = None
    cart_item: Optional[Dict[str, Any]] = None
    total_cost: float
    is_killed: bool
    human_feedback: Optional[str] = None
    created_at: datetime
    updated_at: datetime
    traces: List[AgentTraceOut] = []
    order: Optional[OrderOut] = None
